torch/distributed/_tensor/examples/torchrec_sharding_example.py

```diff
@@ -0,0 +1,173 @@
+
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+
+from torch.distributed._tensor import DTensor, init_device_mesh, Shard
+from torch.distributed._tensor.debug.visualize_sharding import visualize_sharding
+
+
+def backend():
+    return "nccl" if device_type() == "cuda" else "gloo"
+
+
+def device_type():
+    return (
+        "cuda"
+        if torch.cuda.is_available() and torch.cuda.device_count() >= 4
+        else "cpu"
+    )
+
+
+def run_torchrec_row_wise_sharding_example(rank, world_size):
+    # row-wise example:
+    #   one table is sharded by rows within the global ProcessGroup
+    #   In our example, the table's num_embedding is 8, and the embedding dim is 16
+    #   The global ProcessGroup has 4 ranks, so each rank will have one 2 by 16 local
+    #   shard.
+
+    # store each rank of the global ProcessGroup in a tensor
+    global_ranks = torch.arange(world_size)
+    # device mesh is a representation of the worker ranks
+    # create a 1-D device mesh that includes every rank
+    _device_type = device_type()
+    device = torch.device(_device_type)
+    device_mesh = init_device_mesh(device_type=_device_type, mesh_shape=(world_size,))
+
+    # manually create the embedding table's local shards
+    num_embeddings = 8
+    embedding_dim = 16
+    emb_table_shape = torch.Size([num_embeddings, embedding_dim])
+    local_shard_shape = torch.Size(
+        [num_embeddings // world_size, embedding_dim]  # (local_rows, local_cols)
+    )
+    # in our case, the embedding table will be sharded row-wisely into 4 local shards
+    # and each rank will have 1 of them.
+    local_shards = [
+        torch.randn(local_shard_shape, device=device) for _ in range(world_size)
+    ]
+    # row-wise sharding: one shard per rank
+    local_shard = local_shards[rank]
+
+    ###########################################################################
+    # example 1: transform local_shards into DTensor
+    # usage in TorchRec:
+    #   ShardedEmbeddingCollection stores model parallel params in
+    #   _model_parallel_name_to_sharded_tensor which is initialized in
+    #   _initialize_torch_state() and torch.Tensor params are transformed
+    #   into ShardedTensor by ShardedTensor._init_from_local_shards().
+    #
+    #   This allows state_dict() to always return ShardedTensor objects.
+
+    # this is the sharding placement we use in DTensor to represent row-wise sharding
+    # row_wise_sharding_placements means that the global tensor is sharded by first dim
+    # over the 1-d mesh.
+    row_wise_sharding_placements = [Shard(0)]
+    # create a DTensor from the local shard
+    dtensor = DTensor.from_local(
+        local_shard, device_mesh, row_wise_sharding_placements, run_check=False
+    )
+    # display the DTensor's sharding
+    if rank == 0:  # the print result is identical on all ranks
+        print("DTensor's sharding be-like in row-wise sharding:")
+        visualize_sharding(dtensor)
```

Comment:
I think we can enhance visualize_sharding to only do printing on rank 0 by default, instead of doing it on user code.